### ISSUE TITLE
No backward pass for `RmsNorm` if tensor is contiguous

### ISSUE THREAD

**Issue by agerasev** [huggingface/candle#2168]
`RmsNorm` switches to faster implementation if tensor is contiguous:

https://github.com/huggingface/candle/blob/82b641fd2752e3b14db6a9c91faef70e3329f3b5/candle-nn/src/layer_norm.rs#L174-L175

But it does not support backward pass:

https://github.com/huggingface/candle/blob/82b641fd2752e3b14db6a9c91faef70e3329f3b5/candle-nn/src/ops.rs#L640

Maybe it's better to implement `ModuleT` rather than `Module` for `RmsNorm` and use faster implementation only if `train == false`?

**Comment by computer-whisperer** [comment]
Just ran into this, had to make my own copy of LayerNorm as a workaround

**Comment by toolness** [comment]
Same here, I was very confused as to why LayerNorm wasn't working, then I started logging my gradients and noticed that everything before the norm had no gradient.  Very unexpected!

Actually, it looks like the root cause of `LayerNorm`'s problem is different (fixing `RmsNorm` won't automatically fix `LayerNorm`) so I filed #2977 separately.

**Comment by getupforone** [comment]
I also experienced same problem. I spent almost three days to figure out the problem.I just noticed by logging every gradients. The training was not changing weight matrix before layernorm in the Vamap.The root cause of gradient was blocked by layer norm. There were no documentaion about this problem.

**Comment by nxrobins** [comment]
Still reproduces today (2026-05-28) on candle main HEAD \`1d7e9274591ef7eb101d965b71243a41994214d2\`. PR #3526 has been awaiting review since 2026-05-08 with a fix that both removes the contiguous fast-path AND adds a proper \`bwd\` implementation to the \`RmsNorm\` CustomOp2 — including the gradient derivation in the diff. Tested by that PR's author on Qwen2.5-0.5B; independently corroborated by us on the same model.

Minimal standalone reproducer (CPU, <60s, no external assets):
https://github.com/nxrobins/candle-bug-3-rmsnorm-no-bwd

Sibling \`apply_op*_no_bwd\` issues filed this week for the same pattern in other functions:
- #3568 — \`rotary_emb::rope\` / \`rope_i\` / \`rope_thd\`
- #3569 — \`ops::softmax_last_dim\`

Could a maintainer take a look at #3526? It's been ~3 weeks and has a clean fix ready to land.

**Comment by PCfVW** [comment]
Still reproduces on **0.11.0**, the current release, verified 2026-08-01. Rather than open another issue, here is a consolidated index of this cluster, plus four things I do not think have been said in it yet.

**The cluster.** These are all the same `apply_op*_no_bwd` root cause, all open:

| | |
|---|---|
| [#2168](https://github.com/huggingface/candle/issues/2168) | `RmsNorm`, contiguous (this issue) |
| [#2977](https://github.com/huggingface/candle/issues/2977) | `LayerNorm`, contiguous and `remove_mean` |
| [#3011](https://github.com/huggingface/candle/issues/3011) | LayerNorm gradient flow |
| [#3568](https://github.com/huggingface/candle/issues/3568) | `rope`, `rope_i`, `rope_thd` |
| [#3569](https://github.com/huggingface/candle/issues/3569) | `softmax_last_dim` |
| [#3526](https://github.com/huggingface/candle/pull/3526) | PR, fix `RmsNorm` backward |
| [#3612](https://github.com/huggingface/candle/pull/3612) | PR, backward for fused rope |
| [#3613](https://github.com/huggingface/candle/pull/3613) | PR, backward for fused layer_norm |
| [#3724](https://github.com/huggingface/candle/pull/3724) | PR, backward for `softmax_last_dim` |

Cross-linking them is worth doing on its own, because I think the duplication has a mechanical cause: the issues are titled by symptom, one operator each. Someone who hits this through `LayerNorm` searches for "candle layer_norm gradient" and does not find a title that says `RmsNorm`, so they file a new one. @nxrobins is the exception, naming the cause and cross-referencing in both [#3568](https://github.com/huggingface/candle/issues/3568) and [#3569](https://github.com/huggingface/candle/issues/3569).

Worth adding for anyone arriving here: the `Module` wrappers route into the fused ops too, so this is not only about calling `ops::` directly. `RmsNorm::forward` takes the fused path when the input is contiguous (`candle-nn/src/layer_norm.rs:204`), and `LayerNorm::forward` when it is contiguous, `remove_mean` is set and a bias is present (`candle-nn/src/layer_norm.rs:118`). That gating is also why the bug looks
intermittent across codebases.

**1. One site the cluster does not name: `ops::sdpa`.** `candle-nn/src/ops.rs:1317` uses `apply_op3_no_bwd` like the rest. Unlike the norms and `rope`/`rope_i`, it has no `_slow` twin, so there is not even a workaround to point at. I have not run it, so I am flagging it from the source rather than from a measurement.

**2. A systemic alternative to four per-op PRs.** Each open PR adds `bwd()` to one operator, which needs four separate reviews. `candle-core` already exposes the predicate that would settle all of them at the call site, at `candle-core/src/tensor.rs:592`:

```rust
pub fn track_op(&self) -> bool {
    self.is_variable || self.op.is_some()
}
```

Dispatching on that, fused kernel when false, composed form when true, is correct by default, changes no signature and costs inference nothing. [#3568](https://github.com/huggingface/candle/issues/3568) reasoned that "the existence of two variants suggests the no-bwd shape is intentional API design"; from outside it reads more like an omission than a design, since nothing at the call site says the choice affects anything but speed.

If it helps to see it working before deciding, this is exactly what [candle-mi](https://crates.io/crates/candle-mi) has shipped since v0.1.20 (2026-07-27). The three wrappers are one file, about 40 lines of actual logic:

- [`softmax_last_dim`](https://github.com/mi-for-the-rust-of-us/candle-mi/blob/v0.1.20/src/nn_ops.rs#L39)
- [`layer_norm`](https://github.com/mi-for-the-rust-of-us/candle-mi/blob/v0.1.20/src/nn_ops.rs#L65)
- [`rms_norm`](https://github.com/mi-for-the-rust-of-us/candle-mi/blob/v0.1.20/src/nn_ops.rs#L87)
- [the reasoning, in the module preamble](https://github.com/mi-for-the-rust-of-us/candle-mi/blob/v0.1.20/src/nn_ops.rs#L3-L18)

**3. A number, since the four PRs may be stalled on a worry about cost.** I implemented the dispatch downstream and measured it. On my workload the composed forward ran at **0.70x the time of the fused one**, i.e. composed was faster. One model on one GPU, so not a general claim, but it is the only measurement I have seen offered here and it points the opposite way to the intuition.

**4. A test shape that catches the class rather than the instance.** Every report here describes the bug; none proposes a regression test. What worked for me is a count: build the model, run the real forward, `backward()`, then assert that every one of the N parameters received a gradient. Because it is a count it also fails on operators that do not exist yet, which is the property the per-op tests lack. That test is what turned a silent 1-of-29-parameters training run into a loud failure for me, and it is [here](https://github.com/mi-for-the-rust-of-us/candle-mi/blob/v0.1.20/src/diffusion/othello.rs#L808).

Happy to open a PR for the `track_op` dispatch, or for a documentation pass that cross-references the `_slow` twins at each fused function, if either would be welcome.
